### ISSUE TITLE
1cycle never leaves warmup: OneCycleLR gets steps_per_epoch in samples, so every run trains at lrate/25

### ISSUE THREAD

**Issue by thodel** [thodel/serving-atr-inference#96]
`--schedule 1cycle` never leaves its warmup phase. Every kraken run this project has
done has trained at an effectively **constant learning rate of `lrate / 25`**, not at
the rate that was requested.

## Evidence

Read straight out of the checkpoints (`optimizer_states` and `lr_schedulers`):

| run | batch | epoch | global_step | `lr` | `total_steps` |
|---|---|---|---|---|---|
| run 2 (`--lrate 1e-3`) | 256 | 85 | 276,791 | **4.324e-05** | 24,951,540 |
| kraken+ (`--lrate 1e-3`) | 256 | 41 | 136,458 | **4.079e-05** | 24,951,540 |
| run 3 (`--lrate 1e-3`) | 64 | — | — | 4.000e-05 (initial) | 24,951,540 |

`total_steps` is **identical across batch sizes**, and it is not the number of
optimizer steps the run will take:

```
24,951,540 = 30 epochs × 831,718            ← samples in shard_00
     97,470 = 30 epochs ×   3,249           ← actual optimizer steps at batch 256
24,951,540 / 97,470 = 256                   ← exactly the batch size
```

`kraken/train/utils.py` builds the scheduler as

```python
lr_scheduler.OneCycleLR(optim, max_lr=lrate, epochs=epochs,
                        steps_per_epoch=len_train_set, last_epoch=last_epoch)
```

and `len_train_set` is the **sample** count, not the batch count. The cycle is
therefore `batch_size` times too long.

## What that means in practice

`OneCycleLR` starts at `max_lr / div_factor` with the default `div_factor=25`, and
warms up over the first 30% of the cycle — here step 7,485,461. At 3,249 real steps
per epoch that warmup would finish after **2,304 epochs**. No run gets near it:

- run 2 after 85 epochs had moved from 4.000e-05 to 4.324e-05 — an 8% rise across 27
  hours of training.
- The annealing phase, which is the entire point of 1cycle, is never reached.

So `--lrate` behaves as `lrate/25`, roughly constant. Three earlier conclusions have
to be restated:

1. **run 1's blank collapse is fully explained.** `--lrate 1e-4` meant an actual
   4e-6, held there. `docs/TRAINING_PLAN.md` §9a attributes it to step count; the
   learning rate belongs beside it.
2. **"1e-3 works, 1e-4 does not" was really 4e-5 vs 4e-6** — a 10× difference, both
   far below what was asked for.
3. **My description of run 2 "creeping at LR ≈ 0" after epoch 30 was wrong in the
   opposite direction.** The rate was never annealed; it was rising, slowly, the whole
   time. The improvement from 0.68 to 0.78 over 50 epochs happened at ~4e-5.

One thing it does *not* invalidate: run 2 and run 3 both sat at ~4e-5, so the
architecture comparison between them is not confounded by learning rate.

## Options

- **Compensate at the call site.** Pass `--lrate` 25× the intended rate, so the frozen
  start-of-warmup value is what we actually want. Ugly, exact, and needs a comment
  pointing here or nobody will ever understand the number.
- **Use a schedule that is not broken by this.** `constant`, `cosine` or
  `exponential` are stepped the same way but do not encode a total length that has to
  match reality.
- **Divide `epochs` by the batch size** so the cycle length comes out right. Only
  works when `epochs ≥ batch_size`, which for batch 256 it never is.
- **Fix it upstream.** `steps_per_epoch` should be `ceil(len_train_set / batch_size)`.
  Worth reporting to kraken; the workaround is still needed for 7.0.2.

Recommendation: default to `cosine` for new runs, and keep `1cycle` only where a
comment states the 25× compensation explicitly.

## Also worth pinning

`--epochs` does not bound a run when `--quit early` is set — it only sizes the
schedule (the `stage N/∞` in the progress bar). Combined with the above, `--epochs`
currently controls *nothing that happens*: not the stopping point, and not the
learning rate the run actually sees.

**Comment by thodel** [comment]
Behoben in 4c5fa06.

## Zuerst am installierten kraken nachgeprüft

Nicht aus der Doku übernommen, sondern in `.venvs/kraken` auf asterAIx gelesen (kraken **7.0.2**):

```python
# kraken/train/utils.py:295
lr_sched = {'scheduler': lr_scheduler.OneCycleLR(optim, max_lr=lrate,
                                                 epochs=epochs,
                                                 steps_per_epoch=len_train_set, …)}
# kraken/train/vgsl.py:575
len_train_set=len(self.trainer.datamodule.train_set)
```

`len(train_set)` ist die **Sample**-Zahl. Der Befund im Issue steht.

## Was jetzt gilt

Der Default ist `cosine`, und `train_cmd` baut gar keinen 1cycle-Lauf mehr — es gibt auf diesem kraken keine Batchgrösse, bei der der Zeitplan funktioniert, also soll eine Option, die nur ein falsches Ergebnis liefern kann, nicht aus Versehen erreichbar sein. Die Fehlermeldung nennt den Grund und beide brauchbaren Alternativen:

> `--schedule 1cycle is broken on kraken 7.0.2: OneCycleLR is given steps_per_epoch in samples rather than optimizer steps … Use 'cosine' (the default) or 'constant'`

Das Literal bleibt im Typ. Die Job-Records, aus denen alle bisher veröffentlichten Zahlen stammen, tragen `"schedule": "1cycle"`, und ein Params-Modell, das sie nicht mehr parst, würde genau die Läufe unlesbar machen, die belegen, warum die Änderung nötig war.

## Korrigiert, nicht stillschweigend aktualisiert

Von den drei Aussagen, die das Issue zur Richtigstellung auflistet, waren zwei von mir und lagen in **entgegengesetzte** Richtungen falsch:

- `TRAINING_PLAN.md` §9a/§10: run 1 kollabierte nicht, „weil 1cycle die Lernrate bereits weggekühlt hatte" — sie war vom ersten Schritt an bei 4e-6 eingefroren.
- `ARCHITECTURE_SEARCH.md`: run 2 „trainierte 18 von 27 Stunden, nachdem die LR auf null angenähert war" — sie ist die ganze Zeit *gestiegen*, von 4.000e-05 auf 4.324e-05.

Die Schrittzahl-Diagnose (#72) ist davon unberührt und erklärt beide Fälle weiterhin.

## Was das für bestehende Zahlen heisst

Keine gemessene CER ändert sich, aber keine davon lässt sich mit einer künftigen über die Lernrate vergleichen: in jeder Tabelle steht die *angeforderte* Rate, gelaufen ist ein Fünfundzwanzigstel davon. Der Höhen-Sweep aus #91 hat damit bei ~4e-6 statt 1e-4 gemessen — der nächste Sweep ist der erste, bei dem `--lrate` bedeutet, was dasteht.

Bleibt offen und gehört nicht hierher: der Bericht an kraken selbst. `steps_per_epoch` müsste `ceil(len_train_set / batch_size)` sein. Sag Bescheid, wenn ich das Issue upstream aufmachen soll — ich lege nichts in einem fremden Repo an, ohne zu fragen.
